presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1504,6 +1520,83 @@ public PhysicalOperation visitJoin(JoinNode node, LocalExecutionPlanContext cont
+            Expression filterExpression = node.getFilter().get();
+            List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filterExpression);
+            for (FunctionCall spatialFunction : spatialFunctions) {
+                List<Expression> arguments = spatialFunction.getArguments();
+                verify(arguments.size() == 2);
+
+                if (!(arguments.get(0) instanceof SymbolReference) || !(arguments.get(1) instanceof SymbolReference)) {
+                    continue;
+                }
+
+                SymbolReference firstSymbol = (SymbolReference) arguments.get(0);
+                SymbolReference secondSymbol = (SymbolReference) arguments.get(1);
+
+                PlanNode probeNode = node.getLeft();
+                Set<SymbolReference> probeSymbols = getSymbolReferences(probeNode.getOutputSymbols());
+
+                PlanNode buildNode = node.getRight();
+                Set<SymbolReference> buildSymbols = getSymbolReferences(buildNode.getOutputSymbols());
+
+                if (probeSymbols.contains(firstSymbol) && buildSymbols.contains(secondSymbol)) {
+                    return createSpatialLookupJoin(
+                            node,
+                            probeNode,
+                            Symbol.from(firstSymbol),
+                            buildNode,
+                            Symbol.from(secondSymbol),
+                            spatialTest(spatialFunction, true),
+                            removeExpressionFromFilter(filterExpression, spatialFunction),
+                            context);
+                }
+                else if (probeSymbols.contains(secondSymbol) && buildSymbols.contains(firstSymbol)) {
+                    return createSpatialLookupJoin(
+                            node,
+                            probeNode,
+                            Symbol.from(secondSymbol),
+                            buildNode,
+                            Symbol.from(firstSymbol),
+                            spatialTest(spatialFunction, false),
+                            removeExpressionFromFilter(filterExpression, spatialFunction),
+                            context);
+                }
+            }
+
+            throw new VerifyException("No valid spatial relationship found for spatial join");
+        }
+
+        private Optional<Expression> removeExpressionFromFilter(Expression filter, FunctionCall expression)
+        {
+            Expression updatedJoinFilter = replaceExpression(filter, ImmutableMap.of(expression, TRUE_LITERAL));
+            return updatedJoinFilter == TRUE_LITERAL ? Optional.empty() : Optional.of(updatedJoinFilter);
+        }
+
+        private BiPredicate<OGCGeometry, OGCGeometry> spatialTest(FunctionCall functionCall, boolean probeFirst)
+        {
+            switch (functionCall.getName().toString().toLowerCase()) {
+                case ST_CONTAINS:
```

Comment:
Why does this need to be explicitly aware of whether it's a call to ST_CONTAINS or ST_INTERSECTS? We should be able to directly bind to the function described by the function call, as registered in the function registry.
 (i.e., GeoFunctions.stContains and stIntersects)

Reply:
@martint Using functions provided by the plugin requires deserializing Slices for each function invocation. These deserializations happen to be very expensive (see #9890). The cost of point-in-polygon check is proportional to the number of vertexes in the polygon. Deserialization step takes ~40ns per vertex while containment check takes ~10ns per vertex.